### PR TITLE
Fix issue #2292

### PR DESCRIPTION
**Pull request by Filip62**
## What type of PR is this? (check all applicable)

* [ ] Refactor
* [ ] Feature
* [x] Bug Fix
* [ ] Optimization
* [ ] Documentation Update
* [ ] Other: *Replace this with a description of the type of this PR*

## Related Issues & Documents

Fixes [#2292](https://github.com/polybar/polybar/issues/2292)
## Documentation (check all applicable)

* [ ] This PR requires changes to the Wiki documentation (describe the changes)
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [x] Does not require documentation changes

### CODE REVIEW

**Review by patrick96 — CHANGES_REQUESTED** [review]
Thanks for looking into this.

I think this solution only works if all of the indicators are empty (`enabled == 0`)

For example, imagine `m_indicators` had three elements and only the second element was empty, in that case `build` would return true but the output would be:

```
label1__label2
```

(Each underscore represents a call to `builder->space`).
Here, returning false would also not fix this because only trailing spaces are removed by the `get_output` function in `base.inl`.

This is a tricky problem. One solution I could think of is to first iterate over all indicators and add all non-empty labels to a separate vector. We can then iterate over that vector and always add spacing between labels because we know they're not empty.

I think there are also ways to do this in a single iteration but they will probably be a lot more complex than what I described because you have to deal with situations like two consecutive labels being empty.

Also, generally the check for a label being empty is `!label || !*label`, since we know `indicator.second` actually is a valid pointer, you can just use `!*indicator.second` to check if a label is empty.

**Comment by Filip62** [comment]
Actually I think it works.

My initial problem was that a trailing space was added between `<label-layout>` and `<label-indicators>` because even though the indicator was empty, the method `xkeyboard::build` returned `true`, therefore the trailing space was not removed.

Now if at least one of the indicators is not empty only then will `xkeyboard::build` return true, so the space makes sense in that scenario.

I think the calls to `builder->space` work just as expected.
`label-indicator-margin = 1`




2alt

In the current master branch there would be a trailing space at the end of every screenshot.

File: src/modules/xkeyboard.cpp [hunk 1]
```diff
@@ -190,13 +190,17 @@ namespace modules {
       builder->node(m_layout);
     } else if (tag == TAG_LABEL_INDICATOR && !m_indicators.empty()) {
       size_t n{0};
+      size_t enabled{0};
       for (auto&& indicator : m_indicators) {
+        if (!indicator.second) {
```
Comment (patrick96) [inline src/modules/xkeyboard.cpp:195]:
```suggestion
        if (!*indicator.second) {
```

File: src/modules/xkeyboard.cpp [hunk 2]
```diff
@@ -190,13 +190,17 @@ namespace modules {
...
+          enabled++;
+        }
         if (n++) {
           builder->space(m_formatter->get(DEFAULT_FORMAT)->spacing);
         }
         builder->node(indicator.second);
       }
-      return n > 0;
+      return enabled > 0 && n > 0;
```
Comment (patrick96) [inline src/modules/xkeyboard.cpp:203]:
```suggestion
      return enabled > 0;
```

`n` is always greater or equal to `enabled`

**Review by patrick96 — CHANGES_REQUESTED** [review]
I think there was a misunderstading on my part as to what went wrong in your issue.

The extra space(s) after `<label-layout>` can come from one of two places (or both):

* `format-spacing`
* The fact that the format is set to `format = <label-layout> <label-indicator>`

I originally thought your issue was caused by the first, but it isn't because `format-spacing` defaults to 0, so no spacing is added in your case.

You are actually experiencing the second one and you are completely correct that the `build` function returned `true` even though no indicators were present and it should have returned false.

If `format-spacing` is set to anything other than 0, you still get extra spacing. If you want to, you can of course also tackle that, but I think we can just merge the fix for the issue with the wrong return value.

Could you please also add this fix to the `CHANGELOG.md` file. We have just today started keeping a changelog as we merge stuff, there is a new section about in [`CONTRIBUTING.md`](https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog)

Also `!indicator.second` will always return `false` because this only checks if the pointer actually exists and not if the underlying label is empty, that's why you need `*indicator.second`. Right now the `build` function always returns false because `enabled` is never incremented.
That's also why you *never* get any space after the module even though you set `label-indicator-margin = 1`.
margin is always supposed to be added to both sides of a label (or format) and should never be trimmed.

**Comment by Filip62** [comment]
@patrick96 I think I fixed both issues.
From what I tested, everything worked as I would expect.
I hope it works because the solution feels pretty nice with my lame skills. :)

**Review by patrick96 — APPROVED** [review]
Oh wow. I was sure that the solution must be more complex, but this does indeed work. Good job!

**Pull request merged**